### DISCUSSION TITLE
Polymorphism and generics - issues that could perhaps have a solution

### DISCUSSION THREAD

**Discussion by Dryadwoods** [dotnet/roslyn#64602] [Ideas]
Using .NET 6 (code bellow).

Bellow I have a code sample of something that I am trying to accomplish.

 - **Issue 1**:  where we say, use `TTarget` as `User`, we also need to pass the `T` `ID` type.... why by passing `User` is not enought for the compiler to know the `ID` data type? Example: `class UserService : IBaseDBOperation1<User, Guid>` why not `class UserService : IBaseDBOperation1<User>` ?

 - **Issue 2**:  I understand that now we are allowed to have interfaces with methods with code, but why do we still need to define the variable type exactly with both data types and using `var` OR the class name is not enough? Well, we can use `var` or `UserService`, but then the methods are not visible.
 instead of: `IBaseDBOperation1<User, Guid> serviceU1 = new UserService();` ........ `var serviceU2 = new UserService();` ...... this second variable will not see the other methods, the same applies to `UserService serviceU3 = new UserService();`.

**Final note: Everything would be so much easier if C# would allow us to extend a class with more than one other abstract class.... (as of today we are limited to 1).** 

**Objective:** in other words.... avoid copy/paste, and somehow "inject/associate/register/define" more than one "operation class" into a service.... those "operation classes" will be reused a lot in multiple different services.... and I do want to have a "clean/pretty" way of setting this up, but at the same time, the consumer should not worry about "lower/deeper "lever inheritance generics.

Code

    public abstract class BaseDBEntry<T> where T : struct
    {
        protected BaseDBEntry()
        {
            CreatedOn = DateTime.Now;
        }

        public T Id { get; set; }
        public DateTime CreatedOn { get; set; }
        public DateTime? DeletedOn { get; set; }
    }

    public class User : BaseDBEntry<Guid>
    {
        public User() { Id = Guid.NewGuid(); }
        public string Name { get; set; }
    }

    public class Color : BaseDBEntry<long>
    {
        public Color() { Id = DateTime.Now.Ticks; }
        public string RGB { get; set; }
    }

Services

    public interface IBaseDBOperation1<in TTarget, out T> 
                     where TTarget : BaseDBEntry<T> where T : struct
    {
        public bool IsValid(TTarget model) { return true; }

        T GiveMeId(TTarget model) { return model.Id; }
    }

    public interface IBaseDBOperation2<in TTarget, T> 
                     where TTarget : BaseDBEntry<T> where T : struct
    {
        public bool IsValidToDoSomethingElse(TTarget model) { return false; }
    }

    public class UserService : IBaseDBOperation1<User, Guid>, IBaseDBOperation2<User, Guid> { }

    public class ColorService : IBaseDBOperation1<Color, long>, IBaseDBOperation2<Color, long> { }

Consumer

    public class Consumer
    {
        public void Run()
        {
            IBaseDBOperation1<User, Guid> serviceU1 = new UserService();
            IBaseDBOperation2<User, Guid> serviceU2 = new UserService();
            var u = new User { Name = "Carl" };
            var resU1 = serviceU1.IsValid(u);
            var resU2 = serviceU1.GiveMeId(u);
            var resU3 = serviceU2.IsValidToDoSomethingElse(u);

            var serviceU3 = new UserService();
            //serviceU3.XXXXX() --> has no information about the methods we need
            
            UserService serviceU4 = new UserService();
            //serviceU4.XXXXX() --> has no information about the methods we need

            IBaseDBOperation2<Color, long> serviceC1 = new ColorService();
            var c = new Color { RGB = "#FFFFFF" };
            var resC1 = serviceC1.IsValidToDoSomethingElse(c);
        }
    }

    var consumer = new Consumer();
    consumer.Run();

**Comment by jmarolf** [comment]
This might make more sense as a discussion on https://github.com/dotnet/csharplang if what you want it a new language feature. 

When you say
```csharp
//serviceU3.XXXXX() --> has no information about the methods we need
```

Its unclear to me what you expect to be there. Can you replace `XXXXX()` with the actual method you want to call?

**Reply by Dryadwoods** [reply]
When we use `var `or the `Class name` as the object type, then we do not have access to the methods provided by the `BaseDBOperation's`, At the moment we are forced to explicity write IBaseDBOperation1 or IBaseDBOperation2... for that reason we do NOT have access to all the methods the object should have.

Meaning that all the "higher" layers will need to know exactly the lower/deeper lever generics/inheritance types.... and in my opinion, that is a very disadvantage, more when we want to abstract the responsabilities of each.
